### ISSUE TITLE
NGram Tokenizer ignoring search terms smaller than n

### ISSUE THREAD

**Issue by mrdotnic** [elastic/elasticsearch#13710]
NGram Tokenizers are a very useful filter to match parts of compound words. Compound words can be commonly found p.e. in the German language, but occur in English as well (p.e. "Superbowl").
https://www.elastic.co/guide/en/elasticsearch/guide/current/ngrams-compound-words.html

NGram (also called n-gram) Tokenizers split up words into token sequences. Each token has a specified minimum gram length (min_gram) and maximum gram length (max_gram). Tokens consisting of three letters p.e. are called trigrams.
Using trigrams as our starting point, the word "peanut" for example would be divided into the three-letter-tokens "pea", "ean", "anu" and "nut".
Searching for "peanut" would therefore also (partially) match words like "**pea**ce", "S**ean**", "**Anu**bis" and "**nut**rition". They all contain one of these tokens, but might not be words (or the document if they are all in one) we want to see as our results.
Statistically speaking we would then be suffering type I errors (false positives).
https://en.wikipedia.org/wiki/Type_I_and_type_II_errors

Obviously we might then try longer min_gram lengths, p.e. four or five letters long.
With five-grams, "peanut" would be split up in "peanu" and "eanut", giving us far more exact results.
One might come to the conclusion, the longer the n-gram, the better.
Though this is be true to some extent and for some purposes, we will run into another problem.

Elasticsearch's NGram Tokenizers ignore search terms with a length smaller than the specified min_gram!
That means the actual search term will often be significantly shorter and more vague than intended. If there will be a search term left at all.

We might p.e. search for "step by step", trying to find backpacks of this company. Here we have multiple query strings, combined by the AND operator.
If we use our five-gram Tokenizer every word will be ignored. We will have no results resulting in type II errors (false negatives).
Bigrams would respect every word, but lead to a lot of type I errors (see above the "peanut" example).
Trigrams would only exclude the word "by" and respect the rest while being more accurate. But trigrams match the unwanted "Steckdosenleiste" (= multi-outlet power strip). **"Ste"**, "tep", **"Ste"**, "tep" on the one hand and **"ste"**, "tec", "eck", "ckd", "kdo", "dos", "ose", "sen", "enl", "nle", "lei", "eis", "ist", **"ste"** on the other means 50% of the search terms are matched!
So we would might opt for a four-gram Tokenizer, bringing us good results this time.
But then searching for "Jay Smith" would also give us somebody called "John Smith" with the same exact match percentage (type I error).

Even if we wouldn't care for words smaller than four letters, sure enough one can come up with an example where five-grams would be needed.
Searching for "Recht" (= law), a four-gram Tokenizer will match "Echt + rechnung" (= calculation/bill + real):
"Recht" splits up into **"rech"**, **"echt"** while "Echt + rechnung" splits up into **"echt"** and **"rech"**, "echn", "chnu", "hnun", "nung", presenting a perfect match. This might come as a surprise, as those words are in two differents fields (the fields "name" and "description"), making a good example of how unintended and undesired a match can be.
Respecting the token order ("Rech" before "echt") of our search term would help here, but is not supported. Same goes for the "step by step" example.

If it were possible for the Elasticsearch NGram Tokenizer to fall back to smaller sized n-grams, using five-grams for "Smith" and trigrams for "Jay" (instead of ignoring "Jay completely), our problems described above would be gone. We would have relative n-gram precision according to the length of our search terms. But as it stands, we are forced to choose between lots of type I or lots of type II errors.

To sum the examples we have the following situation:
1. A gram size smaller than five leads to problems with our search term "Recht" (type I errors).
2. A gram size larger than three ignores "jay" in "jay smith" (type I error).
3. A gram size larger than four ignores "step by step" (type II error).

This is why a combination of three to five grams (p.e. min_gram=3 and max_gram=5) also fails.

Feel free to play around with the examples in my gist: https://www.found.no/play/gist/6d27486e51bdda88dd84
Edit the min_gram and max_gram sizes to your liking in the "Analysis" section (center left), rerun the queries ("Run" in the top right) and see the altered "Results" section (bottom right) with our Searches # 1, # 2 and # 3.
Searching for "jay smith" corresponds to Search # 1, searching for "step by step" to Search # 2 and finally searching for "recht" to Search # 3. To try out other search terms, edit the queries in the "Searches" section (top right).
Or edit the "Documents" section (top left), to experiment with different words to be searched for.

To summarize: 
If we want to match search terms to compound words ("superbowl"), NGram Tokenizers are a fine tool. They split up words into tokens of a predefined size. 
But Elasticsearch's fixed size NGram Tokenizers force us to choose between either lots of type I or type II errors. What works well for short words like "Jay" won't work for longer words like "Recht" (German for law) and vice versa.
Desired would be a flexible NGram Tokenizer. This means being able to switch from longer NGram Tokenizers (p.e. five-grams) to shorter (p.e. trigrams), as the search terms become shorter. Another, maybe more complicated way could be to make the NGram Tokenizer respect the order of the tokens.

**Comment by clintongormley** [comment]
Hi @mrdotnic 

This is pretty much a duplicate of #9971. In particular see https://github.com/elastic/elasticsearch/issues/9971#issuecomment-89589856. The trick is to use multi-fields with different analysis chains

**Comment by mrdotnic** [comment]
Hi @clintongormley

I just tried the multi fields approach mentioned by @rmuir but without success.

The suggestion works for a search term like "jay" that would be ignored by p.e. a four-gram analyzer but is now considered thanks to the standard analyzer used on top of the four-gram analyzer.
See Search # 2 of the new gist:
https://www.found.no/play/gist/b593300c94524603f73b

However the problem remains when searching for "jay smith" (Search # 1 of the new gist). 
The four-gram analyzer ignores "jay" but keeps "smith", resulting in the unwanted match "john smith".

Maybe I am doing something wrong with my implementation?

**Comment by clintongormley** [comment]
Hi @mrdotnic 

As soon as you stray into the territory of ngrams, you leave exact matches behind you.  Ngrams help you to expand the search to include more potential matches but, for that same reason, you end up with matches which are false positives.

However, the more ngrams that match the better - so the best results should float to the top.  Usually you'd combine this with e.g. `minimum_should_match: 80%` to trim the long tail of poorly matching results.  Combining a query on an ngram field with a query on a full-word (standard analyzer) field is a good way of bumping up the relevance of words that match exactly.

Here's a potential approach that you can use to satisfy your own requirements: make words shorter than your ngrams required matches on the standard field:

    PUT t
    {
      "settings": {
        "analysis": {
          "tokenizer": {
            "4grams": {
              "type": "ngram",
              "min_gram": 4,
              "max_gram": 4,
              "token_chars": [
                "letter",
                "digit"
              ]
            }
          },
          "analyzer": {
            "4grams": {
              "tokenizer": "4grams",
              "filter": [
                "lowercase"
              ]
            },
            "short_std": {
              "tokenizer": "standard",
              "filter": [
                "short",
                "lowercase"
              ]
            }
          },
          "filter": {
            "short": {
              "type": "length",
              "min": 1,
              "max": 3
            }
          }
        }
      },
      "mappings": {
        "t": {
          "properties": {
            "name": {
              "type": "string",
              "analyzer": "standard",
              "fields": {
                "ngrams": {
                  "type": "string",
                  "analyzer": "4grams"
                }
              }
            }
          }
        }
      }
    }
    
    PUT t/t/1
    {"name": "Jay Smith"}
    
    PUT t/t/2
    {"name": "John Smith"}
    
    PUT t/t/3
    {
      "name": "Jay Smithson"
    }
    
    GET /_search
    {
      "query": {
        "bool": {
          "must": [
            {
              "match": {
                "name": {
                  "query": "Jay Smith",
                  "analyzer": "short_std",
                  "operator": "and"
                }
              }
            },
            {
              "match": {
                "name.ngrams": {
                  "query": "Jay Smith",
                  "operator": "or"
                }
              }
            }
          ]
        }
      }
    }

**Comment by felixbarny** [comment]
Thank you very much for the detailed suggestion. I have some trouble with this approach. If you are searching for `smith` only, you don't get any results. Probably because the must query essentially is empty. This is the relevant gist: https://www.found.no/play/gist/2715cdd064250352fd23

One approach would be to merge the two queries into one like this: https://www.found.no/play/gist/64f63fe1fd99f4580ca7
That way I can find all Smiths and Smithsons when searching for 'smith' but searching for 'jay smith' also returns 'john smith' because jay is removed from the .ngram field.

I really think that an optional parameter like `keep_shorter_terms` would solve a lot and since it's optional it can't harm either. But I'm also open for other suggestions.

**Comment by felixbarny** [comment]
any ideas?

**Comment by felixbarny** [comment]
Am I doing something wrong or is that a use case that can't be handled with Elasticsearch? Any chance you can reopen this issue until it gets resolved?

Thanks a lot :)

**Comment by felixbarny** [comment]
Again, this issue is still *not* fixed for me and I hope you reconsider opening it again as I'm still eagerly waiting for a `keep_shorter_terms` option.

**Comment by felixbarny** [comment]
To summarize this issue: Small terms just disappear in the index as well as in the search string. So more documents as expected match because some criteria in the query is just lost. Combining queries on ngram fields in combination with normal fields does not work either. When using an or, I still get more results as expected from the ngram part of the query. When using an and conjunction, I get less results, because the sub words are not matched by the non ngram part of the query. 

So please, reopen this issue.

**Comment by clintongormley** [comment]
@felixbarny using `and`|`or` with ngrams doesn't make much sense... ngrams spread the net wide, and you can use `minimum_should_match` to reduce the long tail of poorly matching documents.  Either way, we use the ngram tokenizer and token filter from lucene, which doesn't have support for this.  If you want to convince Lucene to support it, then I'd open a ticket there.

**Comment by sameerkattel** [comment]
I was also looking for solution to the same problem. And from ES > 6.5 can be achieved this trough [conditional filter](https://www.elastic.co/guide/en/elasticsearch/reference/6.6/analysis-condition-tokenfilter.html)

**Comment by zainabtareen** [comment]
@sameerkattel can you elaborate how it can be used in conjunction with ngram to keep smaller terms, without creating a separate analysis chain?

**Comment by sameerkattel** [comment]
@zainabtareen 

You can set up analyzer with filter and used the analyzer for tokenization 

` {
     "analysis":{
         "filter":{
            "ngrams_filter":{
               "type":"ngram",
               "min_gram":4,
               "max_gram":4
            },
            "ngram_condition":{
               "type":"condition",
               "filter":[
                  "ngrams_filter"
               ],
               "script":{
                  "source":"token.getTerm().length() > 3"
               }
            },
            "edge_grams_filter":{
               "type":"edge_ngram",
               "min_gram":4,
               "max_gram":6
            },
            "edge_ngram_condition":{
               "type":"condition",
               "filter":[
                  "edge_grams_filter"
               ],
               "script":{
                  "source":"token.getTerm().length() > 3"
               }
            }
         },
         "analyzer":{
            "ngrams":{
               "type":"custom",
               "tokenizer":"standard",
               "filter":[
                  "lowercase",
                  "ngram_condition"
               ]
            },
            "edge_grams":{
               "type":"custom",
               "tokenizer":"standard",
               "filter":[
                  "lowercase",
                  "edge_ngram_condition"
               ]
            }
         },
         "normalizer":{
            "lowercase_normalizer":{
               "type":"custom",
               "filter":[
                  "lowercase"
               ]
            }
         }
      }
      }
      `
